src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java

```diff
@@ -48,234 +73,275 @@
+    // prevents grpc on sending watchResponse to a closed watch client.
+    if (this.isClosed()) {
+      return;
+    }
+
+    if (watchResponse.getCreated()) {
+      processCreate(watchResponse);
+    } else if (watchResponse.getCanceled()) {
+      processCanceled(watchResponse);
     } else {
-      // if the watcher is not canceling, cancel it.
-      CompletableFuture<Boolean> completableFuture = this.pendingCancelFutures
-          .get(watchResponse.getWatchId());
-      if (this.pendingCancelFutures.putIfAbsent(watcher.getWatchID(), completableFuture) == null) {
-        cancelWatch(watchResponse.getWatchId());
-      }
+      processEvents(watchResponse);
     }
   }
 
-  /**
-   * resume all the watchers.
-   */
-  private void resumeWatchers(WatcherImpl[] watchers) {
-    for (WatcherImpl watcher : watchers) {
-      if (watcher.callback != null) {
-        watcher.callback.onResuming();
-      }
-      watch(watcher.getKey(), getResumeWatchOptionWithWatcher(watcher), watcher.callback);
+  private synchronized void processError(Throwable t) {
+    // prevents grpc on sending error to a closed watch client.
+    if (this.isClosed()) {
+      return;
+    }
+
+    Status status = Status.fromThrowable(t);
+    if (this.isHaltError(status) || this.isNoLeaderError(status)) {
+      this.notifyWatchers(EtcdExceptionFactory.newEtcdException(status.getDescription()));
+      this.closeGrpcWatchStreamObserver();
+      this.cancelSet.clear();
+      return;
+    }
+    // resume with a delay; avoiding immediate retry on a long connection downtime.
+    scheduledExecutorService.schedule(this::resume, 500, TimeUnit.MILLISECONDS);
+  }
+
+  private synchronized void resume() {
+    this.closeGrpcWatchStreamObserver();
+    this.cancelSet.clear();
+    this.resumeWatchers();
+  }
+
+  // closeGrpcWatchStreamObserver closes the underlying grpc watch stream.
+  private void closeGrpcWatchStreamObserver() {
+    if (this.grpcWatchStreamObserver == null) {
+      return;
+    }
+    this.grpcWatchStreamObserver.onCompleted();
+    this.grpcWatchStreamObserver = null;
+  }
+
+  private boolean isNoLeaderError(Status status) {
+    if (status.getDescription() == null) {
+      return false;
+    }
+    return status.getDescription().contains("no leader");
+  }
+
+  private boolean isHaltError(Status status) {
+    // Unavailable codes mean the system will be right back.
+    // (e.g., can't connect, lost leader)
+    // Treat Internal codes as if something failed, leaving the
+    // system in an inconsistent state, but retrying could make progress.
+    // (e.g., failed in middle of send, corrupted frame)
+    return status.getCode() != Status.Code.UNAVAILABLE && status.getCode() != Status.Code.INTERNAL;
+  }
+
+  private void processCreate(WatchResponse response) {
+    WatcherImpl watcher = this.pendingWatchers.poll();
+
+    this.sendNextWatchCreateRequest();
+
+    if (watcher == null) {
+      // shouldn't happen
```

Comment:
In the case of duplicate create response. this situation can happen but it shouldn't. I think log this case will be fine since correctness is not compromised by duplicate create messages.